muxink/src/demux.rs

```diff
@@ -0,0 +1,254 @@
+///            channel which has messages in the stream doesn't ever take them out.
+pub struct Demultiplexer<S> {
+    stream: Fuse<S>,
+    next_frame: Option<(u8, Bytes)>,
+}
+
+impl<S: Stream> Demultiplexer<S> {
+    /// Creates a new demultiplexer with the given underlying stream.
+    pub fn new(stream: S) -> Demultiplexer<S> {
+        Demultiplexer {
+            stream: stream.fuse(),
+            next_frame: None,
+        }
+    }
+}
+
+impl<S> Demultiplexer<S> {
+    /// Creates a handle listening for frames on the given channel.
+    ///
+    /// Any item on this channel sent to the `Stream` underlying the `Demultiplexer` we used to
+    /// create this handle will be read only when all other messages for other channels have been
+    /// read first. If one has handles on the same channel created via the same underlying
+    /// `Demultiplexer`, each message on that channel will only be received by one of the handles.
+    /// Unless this is desired behavior, this should be avoided.
+    pub fn create_handle(demux: Arc<Mutex<Self>>, channel: u8) -> DemultiplexerHandle<S> {
+        DemultiplexerHandle { channel, demux }
+    }
+}
+
+/// A handle to a demultiplexer.
+///
+/// A handle is bound to a specific channel, see [`Demultiplexer::create_handle`] for details.
+pub struct DemultiplexerHandle<S> {
+    /// Which channel this handle is listening on
+    channel: u8,
+    /// A reference to the underlying demultiplexer.
+    demux: Arc<Mutex<Demultiplexer<S>>>, // (probably?) make sure this is a stdmutex
+}
+
+impl<S> Stream for DemultiplexerHandle<S>
+where
+    S: Stream<Item = Bytes> + Unpin,
+{
+    // TODO Result<Bytes, Error>
+    type Item = Bytes;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Lock the demultiplexer
+        let mut demux = match self.demux.as_ref().try_lock() {
+            Err(_err) => panic!("TODO"), // TODO return Err("TODO")
+            Ok(guard) => guard,
+        };
+
+        // If next_frame has a suitable frame for this channel, return it in a Poll::Ready. If it
+        // has an unsuitable frame, return Poll::Pending. Otherwise, we attempt to read from
+        // the stream.
+        if let Some((ref channel, ref bytes)) = demux.next_frame {
+            if *channel == self.channel {
+                let bytes = bytes.clone();
+                demux.next_frame = None;
+                return Poll::Ready(Some(bytes));
+            } else {
+                return Poll::Pending;
+            }
+        }
+
+        // Try to read from the stream, placing the frame into next_frame and returning
+        // Poll::Pending if its in the wrong channel, otherwise returning it in a Poll::Ready.
+        match demux.stream.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(bytes)) => {
+                let channel: u8 = bytes[0..1][0];
```

Comment:
This will panic when reading 0-sized frames, which I believe an adversary could construct. This needs a custom error variant (compare the `Defragmentizer`'s stream implementation).